lib/BuildSystem/BuildSystem.cpp

```diff
@@ -2262,6 +2262,174 @@ class ArchiveTool : public Tool {
   }
 };
 
+#pragma mark - StaleFileRemovalTool implementation
+
+class StaleFileRemovalCommand : public Command {
+  std::string description;
+
+  std::vector<std::string> expectedOutputs;
+  std::vector<std::string> filesToDelete;
+
+  BuildValue priorValue;
+  bool hasPriorResult = false;
+
+  virtual void configureDescription(const ConfigureContext&, StringRef value) override {
+    description = value;
+  }
+
+  virtual void getShortDescription(SmallVectorImpl<char> &result) override {
+    llvm::raw_svector_ostream(result) << (description.empty() ? "Stale file removal" : description);
+  }
+
+  virtual void getVerboseDescription(SmallVectorImpl<char> &result) override {
+    getShortDescription(result);
+    llvm::raw_svector_ostream(result) << ", stale files: [";
+    for (auto fileToDelete : filesToDelete) {
```

Comment:
Up to you, I think the current code is fine although probably worth a comment towards the subtlety, and please verify there isn't any race condition.